DigitalLearningSolutions.Data.Tests/Services/LoginServiceTests.cs

```diff
@@ -1,239 +1,556 @@
+            GivenDelegateUsersAreVerified(delegateUsers);
+            GivenNoLinkedAccountsFound();
+            GivenNoActiveCentresAreFound();
 
-            //When
-            var (_, returnedDelegateList) = loginService.VerifyUsers(
-                "password",
-                UserTestHelper.GetDefaultAdminUser(),
-                new List<DelegateUser> { delegateUserWithoutPassword }
-            );
+            // When
+            var result = loginService.AttemptLogin(Username, Password);
 
             // Then
-            Assert.IsEmpty(returnedDelegateList);
+            using (new AssertionScope())
+            {
+                result.LoginAttemptResult.Should().Be(LoginAttemptResult.InactiveCentre);
+                result.LogInAdmin.Should().Be(null);
+                result.LogInDelegates.Should().BeEmpty();
+            }
         }
 
         [Test]
-        public void GetVerifiedAdminUserAssociatedWithDelegateUser_Returns_nothing_when_delegate_email_is_empty()
+        public void
+            AttemptLogin_does_not_find_linked_admins_if_verified_admin_already_found_and_no_linked_delegates_and_returns_single_centre_login_result()
         {
             // Given
-            var delegateUserWithoutEmail = UserTestHelper.GetDefaultDelegateUser(emailAddress: null);
+            var adminUser = UserTestHelper.GetDefaultAdminUser(emailAddress: "email@test.com");
+            var adminUsers = new List<AdminUser> { adminUser };
+            GivenAdminUsersAreVerified(adminUsers);
+            GivenResetFailedLoginCountDoesNothing(adminUser);
+            GivenNoLinkedDelegateAccountsFound();
+            AdminUserHasActiveCentre(adminUser);
 
             // When
-            var returnedAdminUser = loginService.GetVerifiedAdminUserAssociatedWithDelegateUser(
-                delegateUserWithoutEmail,
-                "password"
-            );
+            var result = loginService.AttemptLogin(Username, Password);
 
             // Then
-            Assert.IsNull(returnedAdminUser);
+            using (new AssertionScope())
+            {
+                A.CallTo(
+                    () => userVerificationService.GetVerifiedAdminUserAssociatedWithDelegateUsers(
+                        A<List<DelegateUser>>._,
+                        A<string>._
+                    )
+                ).MustNotHaveHappened();
+                result.LoginAttemptResult.Should().Be(LoginAttemptResult.LogIntoSingleCentre);
+                result.LogInAdmin.Should().Be(adminUser);
+                result.LogInDelegates.Should().BeEmpty();
+            }
         }
 
         [Test]
         public void
-            GetVerifiedAdminUserAssociatedWithDelegateUser_Returns_nothing_when_no_admin_account_is_associated_with_delegate()
+            AttemptLogin_does_not_find_linked_admins_if_verified_admin_already_found_but_linked_delegates_are_found_and_returns_single_centre_login_result()

```

Comment:
This test name isn't generally true of our code - if I start with 1 admin and 2 delegates, I'll look for more linked admins, which is the behaviour we want.

I think an improvement would be to replace this test and the previous test with tests for these statements:
* AttemptLogin doesn't look for linked admins if there are no initial, approved, verified delegates.
* AttemptLogin returns single centre login result if 0 or 1 delegates are found

Reply:
The statement was correct for the case admin + delegates going into the account linking step. The call to get more admins only occurred when there was no already an admin (using ??). I've now refactored the private method due to the move of the centre check so that the call will always occur. However we will still always throw away any admin we find if we already have one. (They must be the same admin because of the Email uniqueness and previous checks)